torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,154 @@
+import math
+from numbers import Number
+
+import torch
+from torch.autograd import Variable
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all, lazy_property
+
+
+def _get_batch_shape(bmat, bvec):
+    """
+    Given a batch of matrices and a batch of vectors, compute the combined `batch_shape`.
+    """
+    try:
+        vec_shape = torch._C._infer_size(bvec.shape, bmat.shape[:-1])
+    except RuntimeError:
+        raise ValueError("Incompatible batch shapes: vector {}, matrix {}".format(bvec.shape, bmat.shape))
+    return torch.Size(vec_shape[:-1])
+
+
+def _batch_mv(bmat, bvec):
+    """
+    Performs a batched matrix-vector product, with an arbitrary batch shape.
+    """
+    batch_shape = bvec.shape[:-1]
+    event_dim = bvec.shape[-1]
+    bmat = bmat.expand(batch_shape + (event_dim, event_dim))
+    if batch_shape != bmat.shape[:-2]:
+        raise ValueError("Batch shapes do not match: matrix {}, vector {}".format(bmat.shape, bvec.shape))
+    bvec = bvec.unsqueeze(-1)
+    
+    # conform with `torch.bmm` interface, for matrices with `.dim() == 3`
+    if bvec.dim() == 2:
+        bvec.unsqueeze(0) #_
+     # flatten batch dimensions
+    bvec = bvec.contiguous().view((-1, event_dim, 1))
+    bmat = bmat.contiguous().view((-1, event_dim, event_dim)).expand((bvec.shape[0], -1, -1))
+    return torch.bmm(bmat, bvec).squeeze(-1).view(batch_shape+(event_dim,)) 
+
+
+def _batch_potrf_lower(bmat):
+    """
+    Applies a Cholesky decomposition to all matrices in a batch of arbitrary shape.
+    """
+    n = bmat.size(-1)
+    cholesky = torch.stack([C.potrf(upper=False) for C in bmat.unsqueeze(0).contiguous().view((-1,n,n))])
+    return cholesky.view(bmat.shape)
+
+
+def _batch_diag(bmat):
+    """
+    Returns the diagonals of a batch of square matrices.
+    """
+    n = bmat.size(-1)
+    dims = torch.arange(n, out=bmat.new(n)).long()
+    if isinstance(dims, Variable):
+        dims = dims.data # TODO: why can't I index with a Variable?
+    return bmat[...,dims,dims]
```

Comment:
Wouldn't this work (`size + 1` trick is a nice way to get the diagonal without any copying):
```python
bmat.view(bmat.size(0), -1)[:, ::bmat.size(1) + 1]
```

Reply:
That's totally awesome. Updated.